overlord/snapstate/snapmgr.go

```diff
@@ -509,6 +509,99 @@ func (m *SnapManager) ensureForceDevmodeDropsDevmodeFromState() error {
 	return nil
 }
 
+// changeInFlight returns true if there is any change in the state
+// in non-ready state.
+func changeInFlight(st *state.State) bool {
+	for _, chg := range st.Changes() {
+		if !chg.Status().Ready() {
+			// another change already in motion
+			return true
+		}
+	}
+	return false
+}
+
+// ensureSnapdSnapTransition will migrate systems to use the "snapd" snap
+func (m *SnapManager) ensureSnapdSnapTransition() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	// check if snapd snap is installed
+	var snapst SnapState
+	err := Get(m.state, "snapd", &snapst)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	// nothing to do
+	if snapst.IsInstalled() {
+		return nil
+	}
+
+	// check if the user opts into the snapd snap
+	experimentalAllowSnapd, err := optedIntoSnapdSnap(m.state)
+	if err != nil {
+		return err
+	}
+	// nothing to do: the user does not want the snapd snap yet
+	if !experimentalAllowSnapd {
+		return nil
+	}
+
+	// ensure we only transition systems that have snaps already
+	installedSnaps, err := NumSnaps(m.state)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	// no installed snaps (yet): do nothing (fresh classic install)
+	if installedSnaps == 0 {
+		return nil
+	}
+
+	// get current core snap and use same channel/user for the snapd snap
+	err = Get(m.state, "core", &snapst)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	coreChannel := snapst.Channel
+	userID := snapst.UserID
+
+	if changeInFlight(m.state) {
+		// check that there is no change in flight already, this is a
+		// precaution to ensure the snapd transition is safe
+		return nil
+	}
+
+	// ensure we limit the retries in case something goes wrong
+	var lastSnapdTransitionAttempt time.Time
+	err = m.state.Get("snapd-transition-last-retry-time", &lastSnapdTransitionAttempt)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	now := time.Now()
+	if !lastSnapdTransitionAttempt.IsZero() && lastSnapdTransitionAttempt.Add(6*time.Hour).After(now) {
```

Comment:
we want some jitter here (or we create again a situation where we might hammer the store very regularly)